server/services/notify/notifysubscriptions/notifier.go

```diff
@@ -0,0 +1,250 @@
+// via notifications hints written to the database so that fewer notifications are sent for active
+// blocks.
+type notifier struct {
+	serverRoot string
+	store      Store
+	delivery   SubscriptionDelivery
+	logger     *mlog.Logger
+
+	hints chan *model.NotificationHint
+
+	mux  sync.Mutex
+	done chan struct{}
+}
+
+func newNotifier(params BackendParams) *notifier {
+	return &notifier{
+		serverRoot: params.ServerRoot,
+		store:      params.Store,
+		delivery:   params.Delivery,
+		logger:     params.Logger,
+		done:       nil,
+		hints:      make(chan *model.NotificationHint, 20),
+	}
+}
+
+func (n *notifier) start() {
+	n.mux.Lock()
+	defer n.mux.Unlock()
+
+	if n.done == nil {
+		n.done = make(chan struct{})
+		go n.loop()
+	}
+}
+
+func (n *notifier) stop() {
+	n.mux.Lock()
+	defer n.mux.Unlock()
+
+	if n.done != nil {
+		close(n.done)
+		n.done = nil
+	}
+}
+
+func (n *notifier) loop() {
+	done := n.done
+	var nextNotify time.Time
+
+	for {
+		hint, err := n.store.GetNextNotificationHint(false)
+		switch {
+		case n.store.IsErrNotFound(err):
+			// no hints in table; wait up to an hour or when `onNotifyHint` is called again
+			nextNotify = time.Now().Add(time.Hour * 1)
+			n.logger.Debug("notify loop - no hints in queue", mlog.Time("next_check", nextNotify))
+		case err != nil:
+			// try again in a minute
+			nextNotify = time.Now().Add(time.Minute * 1)
+			n.logger.Error("notify loop - error fetching next notification", mlog.Err(err))
+		case hint.NotifyAt > utils.GetMillis():
+			// next hint is not ready yet; sleep until hint.NotifyAt
+			nextNotify = utils.GetTimeForMillis(hint.NotifyAt)
+		default:
+			// it's time to notify
+			n.notify()
+			continue
+		}
+
+		n.logger.Debug("subscription notifier loop",
+			mlog.Time("next_notify", nextNotify),
+		)
+
+		select {
+		case <-n.hints:
+			// A new hint was added. Wake up and check if next hint is ready to go.
+		case <-time.After(time.Until(nextNotify)):
+			// Next scheduled hint should be ready now.
+		case <-done:
+			return
+		}
+	}
+}
+
+func (n *notifier) onNotifyHint(hint *model.NotificationHint) error {
+	n.logger.Debug("onNotifyHint - enqueing hint", mlog.Any("hint", hint))
+
+	select {
+	case n.hints <- hint:
+	case <-time.After(enqueueNotifyHintTimeout):
+		return errEnqueueNotifyHintTimeout
+	}
+	return nil
+}
+
+func (n *notifier) notify() {
+	var hint *model.NotificationHint
+	var err error
+
+	hint, err = n.store.GetNextNotificationHint(true)
+	if err != nil {
+		// try again later
+		n.logger.Error("notify - error fetching next notification", mlog.Err(err))
+		return
+	}
+
+	if err = n.notifySubscribers(hint); err != nil {
+		n.logger.Error("Error notifying subscribers", mlog.Err(err))
+	}
+}
+
+func (n *notifier) notifySubscribers(hint *model.NotificationHint) error {
+	c := store.Container{
+		WorkspaceID: hint.WorkspaceID,
+	}
+
+	// 	get the subscriber list
+	subs, err := n.store.GetSubscribersForBlock(c, hint.BlockID)
```

Comment:
Should we check the permissions near here to avoid notifying to users about blocks that are not visible for the user?

Reply:
@jespino, do you mean check that the subscriber is still a member of the channel?  Are there other permissions capabilities I'm not thinking of?  My understanding is we are still waiting on a permissions system.

Reply:
I will check for channel membership in `SubscriptionDeliverSlackAttachments` which is called below here.

Reply:
Fixed via https://github.com/mattermost/focalboard/pull/1851/commits/5ec12ade8b06a3799b7201d98b29da3dde8a10dc